ModifierId as String.

catena2w opened the pull request:
fix #178, #208

mike-aksarin commented on src/main/scala/scorex/core/core.scala:
```diff
@@ -20,16 +21,29 @@ package object core {
+    case (Some(f), Some(l)) if f._2 == l._2 => s"[(${f._1},${encoder.encode(f._2)})]"
     case (Some(f), Some(l)) => s"[(${f._1},${encoder.encode(f._2)})..(${l._1},${encoder.encode(l._2)})]"
     case _ => "[]"
   }
```
This could be replaced by:
```
  def idsToString(ids: Seq[(ModifierTypeId, ModifierId)])(implicit enc: ScorexEncoder): String = {
    List(ids.headOption, ids.lastOption)
      .flatten
      .map { case (typeId, id) => s"($typeId,${enc.encode(id)})" }
      .mkString("[", "..", "]")
  }
```

catena2w replied:
fixed

mike-aksarin commented on src/main/scala/scorex/core/utils/ScorexEncoder.scala:
```diff
@@ -0,0 +1,27 @@
+
+}
+
+object ScorexEncoderImpl extends ScorexEncoder
```
Don't we want to rename it to `ScorexEncoder`, similarly to trait name?

catena2w replied:
fixed in a bit different way

catena2w commented on examples/src/main/scala/examples/hybrid/history/HybridHistory.scala:
```diff
@@ -40,10 +40,10 @@ class HybridHistory(val storage: HistoryStorage,
-      case (false, false) => bestPosBlock.parentId sameElements bestPowId
+      case (false, false) => bestPosBlock.parentId == bestPowId
       case (true, false) => ??? //shouldn't be
     }
```
I think it's is quite readable as is and does not corresponds to current PR

mike-aksarin replied:
```
case (false, false)
case (true, false)
```

always look cryptic, but I agree that it's not about this pull request

catena2w commented on examples/src/main/scala/examples/hybrid/history/HybridHistory.scala:
```diff
@@ -93,7 +93,7 @@ class HybridHistory(val storage: HistoryStorage,
 
   override def contains(id: ModifierId): Boolean =
-    if (id sameElements settings.GenesisParentId) true else modifierById(id).isDefined
+    if (id == settings.GenesisParentId) true else modifierById(id).isDefined
```
fixed

catena2w commented on examples/src/main/scala/examples/hybrid/history/HybridHistory.scala:
```diff
@@ -102,15 +102,15 @@ class HybridHistory(val storage: HistoryStorage,
           val mod: ProgressInfo[HybridBlock] = if (isBest) {
             if (isGenesis(powBlock) ||
-              ((powBlock.parentId sameElements bestPowId) && (powBlock.prevPosId sameElements bestPosId))) {
+              ((powBlock.parentId == bestPowId) && (powBlock.prevPosId == bestPosId))) {
```
fixed

kushti commented on src/main/scala/scorex/core/network/message/BasicMessagesRepo.scala:
```diff
@@ -50,18 +50,18 @@ class InvSpec(maxInvObjects: Int) extends MessageSpec[InvData] {
-    require(data._2.size <= maxInvObjects, s"more invs than $maxInvObjects in a message")
-    data._2.foreach(e => require(e.length == NodeViewModifier.ModifierIdSize))
+    require(data._2.lengthCompare(maxInvObjects) <= 0, s"more invs than $maxInvObjects in a message")
+    data._2.foreach(e => require(e.length == 2 * NodeViewModifier.ModifierIdSize))
```
Why this magic-number multiplier?

catena2w commented:
@kushti 
1. Fixed HybridApp in https://github.com/ScorexFoundation/Scorex/pull/294/commits/7987a174ae1d907047eb38cd16ef6e35324ab6ab (not sure it works now in master, I tested this PR by starting Ergo node)
2. Removed magic number in https://github.com/ScorexFoundation/Scorex/pull/294/commits/ba111352025d31c0a8e48e82e6af932af6861af1 (previous version was also correct, while this one is better)

kushti commented:
@catena2w I'm still getting the same:

[ERROR] [SECURITY][07/26/2018 01:48:40.443] [2-Hop-akka.actor.default-dispatcher-7] [akka.actor.ActorSystemImpl(2-Hop)] Uncaught error from thread [2-Hop-akka.actor.default-dispatcher-7]: scorex.core.transaction.state.MinimalState.version()Ljava/lang/String;, shutting down JVM since 'akka.jvm-exit-on-fatal-error' is enabled for ActorSystem[2-Hop]
java.lang.NoSuchMethodError: scorex.core.transaction.state.MinimalState.version()Ljava/lang/String;
	at scorex.core.NodeViewHolder.updateState(NodeViewHolder.scala:231)
	at scorex.core.NodeViewHolder.pmodModify(NodeViewHolder.scala:288)
	at scorex.core.NodeViewHolder.pmodModify$(NodeViewHolder.scala:274)
	at examples.hybrid.HybridNodeViewHolder.pmodModify(HybridNodeViewHolder.scala:21)
	at scorex.core.NodeViewHolder$$anonfun$processLocallyGeneratedModifiers$1.applyOrElse(NodeViewHolder.scala:375)
	at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:171)
	at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:172)
	at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:172)
	at akka.actor.Actor.aroundReceive(Actor.scala:517)
	at akka.actor.Actor.aroundReceive$(Actor.scala:515)
	at examples.hybrid.HybridNodeViewHolder.aroundReceive(HybridNodeViewHolder.scala:21)
